src/hera/v1/cron_workflow.py

```diff
@@ -0,0 +1,56 @@
+"""The implementation of a Hera cron workflow for Argo-based cron workflows"""
+
+from typing import Optional
+
+from argo.workflows.client import (
+    V1alpha1CronWorkflow,
+    V1alpha1CronWorkflowSpec,
+)
+from hera.v1.cron_workflow_service import CronWorkflowService
+from hera.v1.workflow_service import WorkflowService
+from hera.v1.workflow import Workflow
+
+
+class CronWorkflow(Workflow):
+    """A cron workflow representation.
+
+    CronWorkflow are workflows that run on a preset schedule.
+    In essence, CronWorkflow = Workflow + some specific cron options.
+
+    Parameters
+    ----------
+    name: str
+        The cron workflow name. Note that the cron workflow initiation will replace underscores with dashes.
+    service: WorkflowService
+        A workflow service to use for submissions. See `hera.v1.workflow_service.WorkflowService`.
+    schedule: str
+        Schedule at which the Workflow will be run in Cron format. E.g. 5 4 * * *
+    parallelism: int = 50
+        The number of parallel tasks to run in case a task group is executed for multiple tasks.
+    """
+
+    def __init__(
+        self,
+        name: str,
+        schedule: str,
+        service: WorkflowService,
+        cron_service: CronWorkflowService,
+        parallelism: int = 50,
+    ):
+        super().__init__(name, service, parallelism)
+        self.schedule = schedule
+
+        self.cron_workflow_spec = V1alpha1CronWorkflowSpec(schedule=self.schedule, workflow_spec=self.spec)
+        self.cron_workflow = V1alpha1CronWorkflow(metadata=self.metadata, spec=self.cron_workflow_spec)
+        self.cron_service = cron_service
```

Comment:
What do you think about overwriting the `service`, `workflow`, and `spec` properties of the `Workflow`? I think combining properties specific to `Workflow` and `CronWorkflow` (conceptually) into `CronWorkflow` sort of breaks the unique responsibility of one vs. another. If `CronWorkflow` overwrites those properties, then `submit` can maintain the same interface, but execute with a different underlying service, which helps the client avoid potential problems with the submission of the workflow! I don't know the internals of each service (workflow vs. cron workflow) but I assume there are some specific fields different between them that _might_ break functionality if we combine the classes. What do you think? 🙂

Reply:
Interesting point :)
as the dos of argo states: _"In essence, CronWorkflow = Workflow + some specific cron options"_
(https://argoproj.github.io/argo-workflows/cron-workflows/)
That being said - I'm not sure what's the best way of translating that relation into code.

I understand & agree with your comments, but perhaps there is a better way.
How about composition?

Reply:
Ok, very cool! I did not notice that previously, thank you for sharing! One thing I am wary of, though, is allowing the explanation of `CronWorkflow = Workflow + cron options` dictate code patterns. I view it as more of a very accessible explanation of the cron type as I am not sure whether the Argo Workflows code actually reflects that structure.

I see a few options going forward:
- use composition to set the workflow type inside the cron. This requires extracting specific fields, like tasks, off of the workflow type, which adds unnecessary complexity and complicates the cron type a bit
- use inheritance with field overwrites. This maintains a pretty clean interface and avoids some level of code duplication by taking advantage of the provided functionality of the workflow type. This masks away the behavior of the parent workflow type, though, and it might confuse contributors at first as it _might_ be challenging to find exactly which field is used by one over another
- avoid inheritance of workflow by cron and add a base class that acts as the DAG task container, which workflow and cron inherit the `add_task/s` and `add_head/tail` from. This requires a bit of a rewrite that we should probably not do until V2 perhaps.
- avoid inheritance and composition completely. Duplicate the `add_task/s` and `add_head/tail` to cron workflow. Duplication is ok, but when Hera ends up supporting `Step`, for example, it'll need to duplicate the adjustments of `add_step/s` perhaps

I am most in favor of the second option because of verbosity and some savings of code duplication to favor additions of other constructs (future `Step`). In my XP using very verbose implementations, along with some level of code duplication, makes it very easy to pick up something, and my hope is that Hera can showcase that as well. The inheritance, in this case, is quite accessible, and makes it decently clear that some methods are overwritten in order to adjust the used types. Thoughts?

Reply:
Thanks for clearing this up.
I'm not sure though that option #2 is safe.
Perhaps I do not fully understand it, but If the code would go like:

```
class CronWorkflow(Workflow):
    def __init__(
        self,
        name: str,
        schedule: str,
        service: WorkflowService,
        cron_service: CronWorkflowService,
        parallelism: int = 50,
    ):
        super().__init__(name, service, parallelism)
        self.schedule = schedule

        self.workflow = V1alpha1CronWorkflow(metadata=self.metadata, spec=self.spec)
        self.spec = V1alpha1CronWorkflowSpec(schedule=self.schedule, workflow_spec=self.spec)
        self.service = cron_service

    def suspend(self, name: str, namespace: Optional[str] = None):
        self.service.suspend(name, namespace)

    def resume(self, name: str, namespace: Optional[str] = None):
        self.service.resume(name, namespace)

    def submit(self, namespace: Optional[str] = None) -> None:
        self.service.submit(self.workflow, namespace)
```

Then the implementation of `CronWorkflow.add_tasks` would break - since its implementation would refer to `self.spec`. Since it expects it to be a `V1alpha1WorkflowSpec` but got `V1alpha1CronWorkflowSpec` it would crash.

Again, maybe I can't see what you do...
Can u demonstrate your suggestion by code?

Reply:
Ah, yes, you are _definitely_ correct! My apologies for the confusion 🙁  this discussion was extremely valuable! I've tried this myself and ended up going with the 3rd option enumerated above. It makes the code very clean, actually, and as mentioned I think duplication is ok for now until we learn better how to structure these. I keep going back to the explicit vs. implicit and simplicity principles behind Hera and I think the verbosity of an independent cron workflow implementation is very nice! Here's what I wrote and tested in an Argo cluster:

```python
from datetime import datetime, timezone
from typing import Optional
from uuid import uuid4

from argo.workflows.client import *

class CronWorkflowService:

    # TODO: this should go in cron_workflow_service.py
    def __init__(self, domain: str, token: str):
        self.domain = domain
        api_client = Client(Config(domain), token).api_client
        self.service = CronWorkflowServiceApi(api_client=api_client)

    def submit(self, cron_workflow: V1alpha1CronWorkflow, namespace: str = 'default') -> V1alpha1CronWorkflow:
        return self.service.create_cron_workflow(
            namespace, V1alpha1CreateCronWorkflowRequest(cron_workflow=cron_workflow, namespace=namespace)
        )

    def delete(self, name: str, namespace: str = 'default') -> Tuple[object, int, dict]:
        return self.service.delete_cron_workflow(namespace, name)

    def suspend(self, name: str, namespace: str = 'default') -> Tuple[object, int, dict]:
        return self.service.suspend_cron_workflow(
            namespace, name, body=V1alpha1CronWorkflowSuspendRequest(name=name, namespace=namespace)
        )

    def resume(self, name: str, namespace: str = 'default') -> Tuple[object, int, dict]:
        return self.service.resume_cron_workflow(
            namespace, name, body=V1alpha1CronWorkflowResumeRequest(name=name, namespace=namespace)
        )

    def get_cron_workflow_link(self, name: str, namespace: str = 'default') -> str:
        return f'https://{self._domain}/cron-workflows/{namespace}/{name}'


class CronWorkflow:
    # TODO: this should go in cron_workflow.py
    def __init__(
        self,
        name: str,
        schedule: str,
        service: CronWorkflowService,
        parallelism: int = 50,
    ):
        self.name = f'{name.replace("_", "-")}-{str(uuid4()).split("-")[0]}'
        self.schedule = schedule
        self.service = service or CronWorkflowService()
        self.parallelism = parallelism

        self.dag_template = V1alpha1DAGTemplate(tasks=[])
        self.template = V1alpha1Template(name=self.name, steps=[], dag=self.dag_template, parallelism=self.parallelism)
        self.metadata = V1ObjectMeta(name=self.name)
        self.spec = V1alpha1WorkflowSpec(templates=[self.template], entrypoint=self.name)

        self.cron_spec = V1alpha1CronWorkflowSpec(schedule=self.schedule, workflow_spec=self.spec)
        self.workflow = V1alpha1CronWorkflow(
            metadata=self.metadata,
            spec=self.cron_spec,
            status=V1alpha1CronWorkflowStatus(
                active=[], conditions=[], last_scheduled_time=datetime.now(timezone.utc)
            ),
        )

    def add_task(self, t: Task):
        # same implementation as the one in Workflow...
        ...

    def add_tasks(self, *ts: Task):
        # same implementation as the one in Workflow...
        ...

    def add_head(self, t: Task, append: bool = True):
        # same implementation as the one in Workflow...
        ...

    def add_tail(self, t: Task, append: bool = True):
        # same implementation as the one in Workflow...
        ...

    def suspend(self, name: str, namespace: str = 'default'):
        """Suspends the cron workflow"""
        self.service.suspend(name, namespace)

    def resume(self, name: str, namespace: str = 'default'):
        """Resumes execution of the cron workflow"""
        self.service.resume(name, namespace)

    def submit(self, namespace: str = 'default') -> None:
        try:
            self.service.submit(self.workflow, namespace)
        except ValueError as e:
            # the deserialization protocol of the V1alpha1CronWorkflow object creates a V1alpha1CronWorkflowStatus
            # that does not set the `active`, which cannot be `None`. Deserialization happens after this receives 
            # the HTTP response back and it raises a ValueError. This captures that specific error message and simply
            # returns as it's safe for now. Otherwise, it raises the same exception to propagate it to the client
            if str(e) == 'Invalid value for `active`, must not be `None`':
                return
            else:
                raise e
```

What do you think? IMHO, I find this _very_ easy to understand. I makes maintenance and optimization of cron workflows, independently, very easy. It also makes it very clear upon first sight, I think, which should facilitate usage.

Excited to hear what you think! 🙂

Reply:
Honestly, I like more the composition option over the code duplication.
But, as you kindly (😄 ) mentioned above you prefer the duplicates.

It's totally reasonable and respect your decision!

Nevertheless - please allow me to demonstrate just for the sake of it the composition option I mentioned:
```
class CronWorkflow:
    def __init__(
        self,
        name: str,
        schedule: str,
        service: WorkflowService,
        cron_service: CronWorkflowService,
        parallelism: int = 50,
    ):
        self._workflow = Workflow(name, service, parallelism)
        self.schedule = schedule

        self.cron_workflow_spec = V1alpha1CronWorkflowSpec(schedule=self.schedule, workflow_spec=self._workflow.spec)
        self.cron_workflow = V1alpha1CronWorkflow(metadata=self._workflow.metadata, spec=self.cron_workflow_spec)
        self.cron_service = cron_service

    @property
    def workflow(self) -> Workflow:
        return self._workflow

    def suspend(self, name: str, namespace: Optional[str] = None):
        """Suspends the cron workflow"""
        self.cron_service.suspend(name, namespace)

    def resume(self, name: str, namespace: Optional[str] = None):
        """Resumes execution of the cron workflow"""
        self.cron_service.resume(name, namespace)

    def submit(self, namespace: Optional[str] = None) -> None:
        self.cron_service.submit(self.cron_workflow, namespace)

    # These are optional - and matter of taste. User can always type cwf.workflow.add_* ...
    def add_task(self, *_, **__) -> None:
        """proxy method"""
        return self._workflow.add_task(*_, **__)

    def add_tasks(self, *_, **__) -> None:
        """proxy method"""
        return self._workflow.add_tasks(*_, **__)

    def add_head(self, *_, **__) -> None:
        """proxy method"""
        return self._workflow.add_head(*_, **__)

    def add_tail(self, *_, **__) -> None:
        """proxy method"""
        return self._workflow.add_tail(*_, **__)
```

Just to clarify: I'm using your draft as the final version.
Perhaps the snippet above may be useful in the future 🤔 ...
Thank you for listening to my feedback.

Reply:
Thank you so much for illustrating the composition example! It's wonderful. I particularly love the optional proxies. Also, I must say I really appreciate the kindness of your message and understanding 🙂 As we add more feature parity with Argo Workflows, we'll eventually end up building a high level V2 of Hera. I will ping you, if you don't mind, for a discussion on how to organize the new version in the future - I think you'll have great input on the organization of the new version, composition, inheritance patterns, etc.